k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/UidSearchCommand.java

```diff
@@ -0,0 +1,191 @@
+        if (queryString != null) {
+            String encodedQuery = ImapUtility.encodeString(queryString);
+            if (performFullTextSearch) {
+                builder.append("TEXT ");
+            } else {
+                builder.append("OR SUBJECT ").append(encodedQuery).append(" FROM ");
+            }
+            builder.append(encodedQuery).append(" ");
+        }
+    }
+
+    private void addMessageId(StringBuilder builder) {
+        if (messageId != null) {
+            String encodedMessageId = ImapUtility.encodeString(messageId);
+            builder.append("HEADER MESSAGE-ID ").append(encodedMessageId).append(" ");
+        }
+    }
+
+    private void addSince(StringBuilder builder) {
+        if (since != null) {
+            SimpleDateFormat rfc3501DateFormat = new SimpleDateFormat("dd-MMM-yyyy", Locale.US);
+            builder.append("SINCE ").append(rfc3501DateFormat.format(since)).append(" ");
+        }
+    }
+
+    private void addFlags(StringBuilder builder, Set<Flag> flagSet, boolean addNot) {
+        if (flagSet != null) {
+            for (Flag flag : flagSet) {
+                if (addNot) {
+                    builder.append("NOT ");
+                }
+
+                switch (flag) {
+                    case DELETED: {
+                        builder.append("DELETED ");
+                        break;
+                    }
+                    case SEEN: {
+                        builder.append("SEEN ");
+                        break;
+                    }
+                    case ANSWERED: {
+                        builder.append("ANSWERED ");
+                        break;
+                    }
+                    case FLAGGED: {
+                        builder.append("FLAGGED ");
+                        break;
+                    }
+                    case DRAFT: {
+                        builder.append("DRAFT ");
+                        break;
+                    }
+                    case RECENT: {
+                        builder.append("RECENT ");
+                        break;
+                    }
+                    default: {
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    Builder newBuilder() {
+        return new Builder()
+                .useUids(useUids)
+                .queryString(queryString)
+                .messageId(messageId)
+                .performFullTextSearch(performFullTextSearch)
+                .since(since)
+                .requiredFlags(requiredFlags)
+                .forbiddenFlags(forbiddenFlags)
+                .listener(listener);
+    }
+
+    public static class Builder extends FolderSelectedStateCommand.Builder<UidSearchCommand, Builder> {
+
+        public Builder useUids(boolean useUids) {
+            command.useUids = useUids;
+            return builder;
+        }
+
+        public Builder queryString(String queryString) {
```

Comment:
so much boilerplate :\ would this be easier to do with [AutoValue](https://github.com/google/auto/blob/master/value/userguide/index.md)?